pallets/delegating/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+                Err(Error::<T>::SelectNoValidator)?
+            }
+
+            // check if controller has call delegated
+            if <DelegatedToValidators<T>>::contains_key(controller.clone()){
+                Err(Error::<T>::AlreadyDelegated)?
+            }
+
+            // check target validators in candidate_validators
+            let candidate_validators = <CandidateValidators<T>>::get().unwrap();
+            for validator in validators.clone() {
+                if !candidate_validators.contains(&validator){
+                    error!("Validator AccountId  isn't in candidateValidators");
+                    Err(Error::<T>::NotInCandidateValidator)?
+                }
+            }
+
+            // get avg score to validators
+            let validators_vec = Self::cut_credit_score(controller.clone(), validators.clone());
+            Self::_delegate(controller.clone(), validators_vec);
+
+            let credit_delegate_info = CreditDelegateInfo{
+                delegator: controller.clone(),
+                score: T::CreditInterface::get_credit_score(controller.clone()).unwrap(),
+                validators: validators.clone(),
+            };
+            <DelegatedToValidators<T>>::insert(controller.clone(), credit_delegate_info);
+
+            Self::deposit_event(RawEvent::Delegated(controller));
+            Ok(())
+        }
+
+        #[weight = 10_000]
+        pub fn undelegate(origin) -> dispatch::DispatchResult {
+            let controller = ensure_signed(origin)?;
+
+            if !<DelegatedToValidators<T>>::contains_key(controller.clone()){
+                Err(Error::<T>::NotDelegate)?
+            }
+            Self::_undelegate(controller.clone());
+            <DelegatedToValidators<T>>::remove(controller.clone());
+
+            Self::deposit_event(RawEvent::UnDelegated(controller));
+            Ok(())
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    fn _delegate(delegator: T::AccountId, validators_vec: Vec<(T::AccountId, u64)>) {
+        for (validator, score) in validators_vec {
+            let has_delegated =
+                <HasDelegatedToValidator<T>>::get(delegator.clone(), validator.clone())
+                    .unwrap_or(false);
+            if has_delegated == false {
+                // delegate first times
+                if CandidateDelegators::<T>::contains_key(validator.clone()) {
+                    let mut delegators = CandidateDelegators::<T>::take(validator.clone());
+                    delegators.push((delegator.clone(), score));
+                    CandidateDelegators::<T>::insert(validator.clone(), delegators);
+                } else {
+                    let delegators = vec![(delegator.clone(), score)];
+                    CandidateDelegators::<T>::insert(validator.clone(), delegators);
+                }
+            } else {
+                // delegated, update score
+                let delegators = CandidateDelegators::<T>::take(validator.clone());
```

Comment:
how this case will happen? If the validator is already delegated, it will not pass the  <DelegatedToValidators<T>>::contains_key(controller.clone) check. I am missing something here?

Reply:
unless the vector of validators the delegator chosen can contains repeated validators. then this will happen. But in this case, the update score is not added up, just replaced by new score.

Reply:
Line 285, In function check_and_adjust_delegated_score(), we will update delegators' delegated credit score every era.